### PR TITLE
prune: scan and unlink already pruned block files on startup

### PR DESCRIPTION
**Pull request by andrewtoth**
There are a few cases where we can mark a block and undo file as pruned in our block index, but not actually remove the files from disk.
1. If we call `FindFilesToPrune` or `FindFilesToPruneManual` and crash before `UnlinkPrunedFiles`.
2. If on Windows there is an open file handle to the file somewhere else when calling `fs::remove` in `UnlinkPrunedFiles` (https://en.cppreference.com/w/cpp/filesystem/remove, https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-deletefilew#remarks). This could be from another process, or if we are calling `ReadBlockFromDisk`/`ReadRawBlockFromDisk` without having a lock on `cs_main` (which has been allowed since https://github.com/bitcoin/bitcoin/commit/ccd8ef65f93ed82a87cee634660bed3ac17d9eb5).

This PR mitigates this by scanning all pruned block files on startup after `LoadBlockIndexDB` and unlinking them again.

### CODE REVIEW

**Review by hernanmarino — COMMENTED** [review]
Concept ACK

**Review by pablomartin4btc — COMMENTED** [review]
cr & tested ACK. I've also checked a bit of the history if your proposal in all these PRs #26308,  #26316 (closed in favour of this current one), #17161 and #25232.

File: src/node/blockstorage.cpp [hunk 1]
```diff
@@ -371,6 +371,28 @@ bool BlockManager::LoadBlockIndexDB(const Consensus::Params& consensus_params)
...
+void BlockManager::ScanAndUnlinkAlreadyPrunedFiles() {
+    AssertLockHeld(::cs_main);
+    if (!m_have_pruned) {
+        return;
+    }
+
+    std::set<int> block_files_to_prune;
+    for (int file_number = 0; file_number < m_last_blockfile; file_number++) {
+        if (m_blockfile_info[file_number].nSize == 0
+            && !CAutoFile(OpenBlockFile(FlatFilePos(file_number, 0), true), SER_DISK, CLIENT_VERSION).IsNull()) {
```
Comment (luke-jr) [inline src/node/blockstorage.cpp:383]: This seems likely to be somewhat I/O-heavy.

Can't we just call unlink and ignore any errors?

File: src/test/blockmanager_tests.cpp [hunk 2]
```diff
@@ -39,4 +41,35 @@ BOOST_AUTO_TEST_CASE(blockmanager_find_block_pos)
...
+    // Check that the file is unlinked after ScanAndUnlinkAlreadyPrunedFiles
+    WITH_LOCK(Assert(m_node.chainman)->GetMutex(), m_node.chainman->m_blockman.ScanAndUnlinkAlreadyPrunedFiles());
+    BOOST_CHECK(CAutoFile(OpenBlockFile(pos, true), SER_DISK, CLIENT_VERSION).IsNull());
+
+    // Check that the new tip file has not been removed
+    const CBlockIndex* new_tip{WITH_LOCK(Assert(m_node.chainman)->GetMutex(), return m_node.chainman->ActiveChain().Tip())};
+    BOOST_CHECK_NE(old_tip, new_tip);
+    const int new_file_number{WITH_LOCK(Assert(m_node.chainman)->GetMutex(), return new_tip->GetBlockPos().nFile)};
+    const FlatFilePos new_pos(new_file_number, 0);
+    BOOST_CHECK(!CAutoFile(OpenBlockFile(new_pos, true), SER_DISK, CLIENT_VERSION).IsNull());
```
Comment (MarcoFalke) [inline src/test/blockmanager_tests.cpp:72]: Same

**Review by maflcko — APPROVED** [review]
Nice. Concept ACK

File: src/test/blockmanager_tests.cpp [hunk 3]
```diff
@@ -39,4 +41,43 @@ BOOST_AUTO_TEST_CASE(blockmanager_find_block_pos)
     BOOST_CHECK_EQUAL(actual.nPos, BLOCK_SERIALIZATION_HEADER_SIZE + ::GetSerializeSize(params->GenesisBlock(), CLIENT_VERSION) + BLOCK_SERIALIZATION_HEADER_SIZE);
 }
 
+BOOST_FIXTURE_TEST_CASE(blockmanager_scan_unlink_already_pruned_files, TestChain100Setup)
+{
+    // Cap last block file size, and mine new block in a new block file.
+    const CBlockIndex* old_tip{WITH_LOCK(Assert(m_node.chainman)->GetMutex(), return m_node.chainman->ActiveChain().Tip())};
```
Comment (MarcoFalke) [inline src/test/blockmanager_tests.cpp:47]:
```suggestion
    const auto& chainman{Assert(m_node.chainman)};
    const CBlockIndex* old_tip{WITH_LOCK(chainman.GetMutex(), return chainman.ActiveTip())};
```

nit, if you retouch: Could use an alias to reduce verbosity

**Review by maflcko — COMMENTED** [review]
Feel free to ignore the nit

**Review by pablomartin4btc — COMMENTED** [review]
re-ACK with added functional test 3141eab9c669488a2e7fef5f60d356ac92294922.

**Review by furszy — APPROVED** [review]
Code review ACK 3141eab9
Left a non-blocking nit. No need to do it.

**Review by theStack — COMMENTED** [review]
Concept ACK

Looks good to me, planning to do a more thorough review tomorrow. A (non-blocking) suggestion on how to write the functional test shorter. In particular, the `sync_blocks` call is basically a no-op if we only have one node:
```diff
diff --git a/test/functional/feature_remove_pruned_files_on_startup.py b/test/functional/feature_remove_pruned_files_on_startup.py
index ca0e5ace9..5df101863 100755
--- a/test/functional/feature_remove_pruned_files_on_startup.py
+++ b/test/functional/feature_remove_pruned_files_on_startup.py
@@ -13,11 +13,9 @@ class FeatureRemovePrunedFilesOnStartupTest(BitcoinTestFramework):
         self.extra_args = [["-fastprune", "-prune=1"]]
 
     def mine_batches(self, blocks):
-        n = blocks // 250
-        for _ in range(n):
+        for _ in range(blocks // 250):
             self.generate(self.nodes[0], 250)
         self.generate(self.nodes[0], blocks % 250)
-        self.sync_blocks()
 
     def run_test(self):
         blk0 = os.path.join(self.nodes[0].datadir, self.nodes[0].chain, 'blocks', 'blk00000.dat')
@@ -25,10 +23,8 @@ class FeatureRemovePrunedFilesOnStartupTest(BitcoinTestFramework):
         blk1 = os.path.join(self.nodes[0].datadir, self.nodes[0].chain, 'blocks', 'blk00001.dat')
         rev1 = os.path.join(self.nodes[0].datadir, self.nodes[0].chain, 'blocks', 'rev00001.dat')
         self.mine_batches(800)
-        fo1 = os.open(blk0, os.O_RDONLY)
-        fo2 = os.open(rev1, os.O_RDONLY)
-        fd1 = os.fdopen(fo1)
-        fd2 = os.fdopen(fo2)
+        fd1 = os.fdopen(os.open(blk0, os.O_RDONLY))
+        fd2 = os.fdopen(os.open(rev1, os.O_RDONLY))
         self.nodes[0].pruneblockchain(600)
 
         # Windows systems will not remove files with an open fd
```

**Review by theStack — APPROVED** [review]
Code-review ACK 3141eab9c669488a2e7fef5f60d356ac92294922

FWIW, was curious how calling `fs::remove` on thousands of non-existing files would affect the performance (on mainnet, there is currently >3400 .blk/.rev files). But as expected, it's not a big deal:
```
$ cat remove_test.cpp                            
#include <filesystem>
#include <string>
int main() {
    for (int i = 0; i < 10000; ++i) {
        std::string filename = "/home/thestack/" + std::to_string(i) + ".dat";
        std::filesystem::remove(filename);
    }
}
$ c++ -std=c++17 -o remove_test remove_test.cpp  
$ time ./remove_test                                                                                                        
    0m00.08s real     0m00.01s user     0m00.07s system

```

**Comment by achow101** [comment]
ACK 3141eab9c669488a2e7fef5f60d356ac92294922

**Pull request merged**